fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByRecords.java

```diff
@@ -0,0 +1,677 @@
+     * @param start the (inclusive) beginning primary key of the range to build (or <code>null</code> to start from the beginning)
+     * @param end the (exclusive) end primary key of the range to build (or <code>null</code> to go to the end)
+     * @return a future with the key of the first record not processed by this range rebuild
+     * @throws OnlineIndexer.RecordBuiltRangeException if the given range contains keys already processed by the index build
+     */
+    @Nonnull
+    public CompletableFuture<Key.Evaluated> buildUnbuiltRange(@Nonnull FDBRecordStore store,
+                                                              @Nullable Key.Evaluated start,
+                                                              @Nullable Key.Evaluated end) {
+        return buildUnbuiltRange(store, start, end, null);
+    }
+
+    // just like the overload that doesn't take a recordsScanned
+    @Nonnull
+    private CompletableFuture<Key.Evaluated> buildUnbuiltRange(@Nonnull FDBRecordStore store,
+                                                               @Nullable Key.Evaluated start, @Nullable Key.Evaluated end,
+                                                               @Nullable AtomicLong recordsScanned) {
+        return buildUnbuiltRange(store, convertOrNull(start), convertOrNull(end), recordsScanned)
+                .thenApply(tuple -> (tuple == null) ? null : Key.Evaluated.fromTuple(tuple));
+    }
+
+    // Helper function with the same behavior as buildUnbuiltRange, but it works on tuples instead of primary keys.
+    @Nonnull
+    private CompletableFuture<Tuple> buildUnbuiltRange(@Nullable Tuple start, @Nullable Tuple end) {
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "buildUnbuiltRange",
+                LogMessageKeys.RANGE_START, start,
+                LogMessageKeys.RANGE_END, end);
+
+        return buildCommitRetryAsync((store, recordsScanned) -> buildUnbuiltRange(store, start, end, recordsScanned),
+                true,
+                additionalLogMessageKeyValues);
+    }
+
+    @VisibleForTesting
+    @Nonnull
+    CompletableFuture<Key.Evaluated> buildUnbuiltRange(@Nullable Key.Evaluated start, @Nullable Key.Evaluated end) {
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "buildUnbuiltRange",
+                LogMessageKeys.RANGE_START, start,
+                LogMessageKeys.RANGE_END, end);
+        return buildCommitRetryAsync((store, recordsScanned) -> buildUnbuiltRange(store, start, end, recordsScanned),
+                true,
+                additionalLogMessageKeyValues);
+    }
+
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    private RecordCursorResult<FDBStoredRecord<Message>> castCursorResult(RecordCursorResult<?> result) {
+        if (result == null) {
+            throw new MetaDataException("Unexpected null result");
```

Comment:
It looks like this is still a `MetaDataException`, when it shouldn't be. I also think the signature on this method can be improved so that it takes a `RecordCursorResult<FDBStoredRecord<Message>>` with the improved generics on `iterateRangeOnly`

Reply:
Oops, missed this one (change in IndexingByIndex only).